### ISSUE TITLE
Illegal operation on const operands does not error at compile-time.

### ISSUE THREAD

**Issue by Boiethios** [rust-lang/rust#54321]
Const errors should be catched at compile-time.

I tried this code:

```rust
fn main() {
    println!("{}", size_of::<[(); 3]>() / size_of::<()>());
}

```
I expected to have a compile-time error, like in this code:

```rust
const ERROR: usize = size_of::<[(); 3]>() / size_of::<()>();
```

```
  |
3 | const ERROR: usize = size_of::<[(); 3]>() / size_of::<()>();
  | ^^^^^^^^^^^^^^^^^^^^^--------------------------------------^
  |                      |
  |                      attempt to divide by zero
  |

```

Instead, it compiles fine and panics when it is run.

**Comment by Centril** [comment]
cc @varkor @oli-obk @RalfJung

**Comment by Centril** [comment]
Actually... this seems like expected behavior...?

Perhaps a warning because the compiler can constant fold it and see the problem, but you haven't forced CTFE as with `const ERROR ...`.

**Comment by RalfJung** [comment]
This is deliberate and not a bug. In fact we have an RFC saying we must not make this a bug (can't find it right now) -- because if we did, making the analysis smarter would make more code error.

**Comment by RalfJung** [comment]
Yeah, one could argue that `transform/const_prop.rs` should lint about this, like it lints about `1/0` (that's actually a deny-by-default lint). But TBH that analysis is already quite horrible, I'd rather not make it even more horrible.

**Comment by RalfJung** [comment]
@Boiethios The key difference between `const ERROR` and `println` is that with `const ERROR`, you are explicitly asking the compiler "evaluate this at compile-time" -- so we have no choice but error about anything that goes wrong.

This is extremely different from writing code in a normal function body that consists entirely of constants, so the compiler *could* evaluate at compile-time if it wanted -- but it doesn't have to, doing so is just an optimization.

Essentially, this is the difference between compile-time function evaluation and constant propagation, which happen in totally different compilation phases and under different circumstances.

**Comment by nagisa** [comment]
Closing as expected behaviour. I also mildly recall a discussion about this going towards not making this an error.

**Comment by RalfJung** [comment]
> I hope that there are some plans to make const evaluation more widely available.

There are -- by making more features available for const evaluation. However, 

> an operation between 2 const operands should also be const

is explicitly something we do NOT want to happen. Whether something is const-eval'ed or not is observable, and the programmer should ask for that explicitly.
